fastlane/lib/fastlane/helper/sh_helper.rb

```diff
@@ -2,30 +2,64 @@
 
 module Fastlane
   module Actions
-    # Execute a shell command
-    # This method will output the string and execute it
-    # Just an alias for sh_no_action
-    # When running this in tests, it will return the actual command instead of executing it
-    # @param log [Boolean] should fastlane print out the executed command
-    # @param error_callback [Block] a callback invoked with the command output if there is a non-zero exit status
+    # Execute a shell command. This method will output the string and execute it.
+    #
+    # The `log` argument accepts either a boolean or an aray of symbols
+    # determining what should be printed and what should be suppressed:
+    #
+    # ```
+    # # $ echo foo && exit 1
+    # sh("echo foo && exit 1", log: [:command])
+    #
+    # # > foo
+    # sh("echo foo && exit 1", log: [:output])
+    #
+    # # Exit status was 1 instead of 0.
+    # # foo
+    # sh("echo foo && exit 1", log: [:error])
+    # ```
+    #
+    # You can combine these together. The argument also supports just `true`
+    # which enables all output and `false` which suppresses all output.
+    #
+    # When running this in tests, it will return the actual command instead of
+    # executing it.
+    #
+    # @param log [Boolean, Array] Should fastlane print out the executed command, its output and error
+    # @param error_callback [Block] A callback invoked with the command output if there is a non-zero exit status
     def self.sh(*command, log: true, error_callback: nil, &b)
-      sh_control_output(*command, print_command: log, print_command_output: log, error_callback: error_callback, &b)
+      parse_log_argument = proc { |key|
+        case log
+        when true, false then log
+        when Array then log.include?(key)
+        end
+      }
+      sh_control_output(
+        *command,
+        print_command: parse_log_argument.call(:command),
+        print_command_output: parse_log_argument.call(:output),
+        print_command_error: parse_log_argument.call(:error),
+        error_callback: error_callback,
+        &b
+      )
     end
 
+    # Just an alias for `sh`.
     def self.sh_no_action(*command, log: true, error_callback: nil, &b)
-      sh_control_output(*command, print_command: log, print_command_output: log, error_callback: error_callback, &b)
+      sh(*command, log: log, error_callback: error_callback, &b)
```

Comment:
why the change here?

Reply:
As `sh_no_action` is supposed to be an alias for `sh` (they had exactly the same implementations), I didn't want to duplicate the code in `sh` and `sh_no_action`. So I made `sh_no_action` call `sh`.

If you feel that this added level of indirection is bad, I can copy&paste the changed implementation from `sh` here as well.

Reply:
Nah, I was just wondering why the change is needed now. Wasn't this the case before as well?

Reply:
got it.